File: packages/testkit/src/scenarios.ts
/**
 * High-level scenario scripting (mirrors `fusionkit_testkit.scenarios`).
 *
 * Encodes FusionKit's fusion call graph — panel fanout, then judge analysis,
 * then synthesizer answer, with judge and synthesizer often the same endpoint
 * consumed FIFO — so a test scripts a whole fused turn in one call.
 */

import { asBehavior } from "./behaviors.js";
import type { SimBehaviorInput } from "./behaviors.js";
import type { ProviderSimHandle } from "./provider-sim.js";

/** A well-formed judge analysis JSON reply (the judge model's first fuse-step turn). */
export function judgeAnalysis(overrides: {
  consensus?: string[];
  contradictions?: string[];
  unique_insights?: string[];
  coverage_gaps?: string[];
  likely_errors?: string[];
  recommended_final_structure?: string[];
  best_trajectory?: string;
} = {}): string {
  return JSON.stringify({
    consensus: overrides.consensus ?? ["candidates agree"],
    contradictions: overrides.contradictions ?? [],
    unique_insights: overrides.unique_insights ?? [],
    coverage_gaps: overrides.coverage_gaps ?? [],
    likely_errors: overrides.likely_errors ?? [],
    recommended_final_structure: overrides.recommended_final_structure ?? [],
    ...(overrides.best_trajectory !== undefined ? { best_trajectory: overrides.best_trajectory } : {})
  });
}

export type FusedTurnScript = {
  /** Provider model name -> panel candidate reply (what the journal records). */
  candidates: Record<string, SimBehaviorInput>;
  /** The judge endpoint's provider model name. */
  judgeModel: string;
  /** The synthesizer's final answer (second judge-endpoint turn by default). */
  answer: SimBehaviorInput;
  /** Judge analysis JSON; defaults to a well-formed {@link judgeAnalysis}. */
  analysis?: string;
  /** Set when the synthesizer is a different endpoint than the judge. */
  synthesizerModel?: string;
};

/** Script one full fused turn: panel candidates, judge analysis, synthesis. */
export async function scriptFusedTurn(sim: ProviderSimHandle, script: FusedTurnScript): Promise<void> {
  for (const [model, reply] of Object.entries(script.candidates)) {
    await sim.queue(model, [reply]);
  }
  const analysis = { reply: script.analysis ?? judgeAnalysis() };
  if (script.synthesizerModel === undefined || script.synthesizerModel === script.judgeModel) {
    await sim.queue(script.judgeModel, [analysis, asBehavior(script.answer)]);
    return;
  }
  await sim.queue(script.judgeModel, [analysis]);
  await sim.queue(script.synthesizerModel, [asBehavior(script.answer)]);
}

File: python/fusionkit-testkit/src/fusionkit_testkit/scenarios.py
"""High-level scenario scripting.

These helpers encode FusionKit's fusion call graph (panel fanout, then judge
analysis, then synthesizer answer — with judge and synthesizer often being the
same endpoint, consumed FIFO) so a test scripts a whole fused turn in one
call instead of re-deriving the call order every time.
"""

from __future__ import annotations

import json
from collections.abc import Mapping

from fusionkit_testkit.behaviors import Behavior
from fusionkit_testkit.server import RouteKitSimulator

ScriptedReply = str | Behavior


def as_behavior(value: ScriptedReply) -> Behavior:
    return value if isinstance(value, Behavior) else Behavior(reply=value)


def judge_analysis(
    *,
    consensus: list[str] | None = None,
    contradictions: list[str] | None = None,
    unique_insights: list[str] | None = None,
    coverage_gaps: list[str] | None = None,
    likely_errors: list[str] | None = None,
    recommended_final_structure: list[str] | None = None,
    best_trajectory: str | None = None,
) -> str:
    """A well-formed judge analysis JSON reply (the judge model's first turn)."""
    payload: dict[str, object] = {
        "consensus": consensus or ["candidates agree"],
        "contradictions": contradictions or [],
        "unique_insights": unique_insights or [],
        "coverage_gaps": coverage_gaps or [],
        "likely_errors": likely_errors or [],
        "recommended_final_structure": recommended_final_structure or [],
    }
    if best_trajectory is not None:
        payload["best_trajectory"] = best_trajectory
    return json.dumps(payload)


def script_fused_turn(
    sim: RouteKitSimulator,
    *,
    candidates: Mapping[str, ScriptedReply],
    judge_model: str,
    answer: ScriptedReply,
    analysis: str | None = None,
    synthesizer_model: str | None = None,
) -> None:
    """Script one full fused turn: panel candidates, judge analysis, synthesis.

    ``candidates`` maps namespaced RouteKit model ids (what the simulator
    journals) to their panel replies. When ``synthesizer_model`` is unset the
    judge model serves both fuse-step roles, consumed in order: analysis first,
    answer second.
    """
    for model, reply in candidates.items():
        sim.queue(model, as_behavior(reply))
    analysis_behavior = Behavior(reply=analysis if analysis is not None else judge_analysis())
    if synthesizer_model is None or synthesizer_model == judge_model:
        sim.queue(judge_model, analysis_behavior, as_behavior(answer))
    else:
        sim.queue(judge_model, analysis_behavior)
        sim.queue(synthesizer_model, as_behavior(answer))
